Updating advanced git chapter with Zipf's content (#349)
merely-useful/merely-useful.github.io: py-rse/git-advanced.Rmd

a mathematical optimization problem,
for which we can use the `scipy.optimize` library.
Following @Moreno2016,
we use Brent's method with \(1 < \beta \leq 4\).

```python
from scipy.optimize import minimize_scalar

def get_power_law_params(word_counts):
    """Get the power law parameters."""
    mle = minimize_scalar(nlog_likelihood, bracket=(1 + 1e-10, 4),
                          args=(word_counts), method='brent')
    beta = mle.x
    alpha = 1 / (beta - 1)
    return alpha, beta
```

We can then plot the fitted curve on the plot axes (`ax`)
defined in the `plotcounts.py` script,

```python
def plot_fit(curve_xmin, curve_xmax, max_rank, beta, ax):
    """
    Plot the power law curve that was fitted to the data.

    Parameters
    ----------
    curve_xmin : float
        Minimum x-bound for fitted curve
    curve_xmax : float
        Maximum x-bound for fitted curve
    max_rank : int
        Maximum word frequency rank.
    beta : float
        Estimated beta parameter for the power law.
    ax : matplotlib axes
        Scatter plot to which the power curve will be added.
    """
    xvals = np.arange(curve_xmin, curve_xmax)

>     yvals = max_rank * (xvals**(-beta + 1))

Reviewer: Maybe we should rewrite this to use `alpha` not `beta`, so it links better with the previous section.

Author: That's probably a good idea. I guess we can wait until you resolve #367 to rewrite, since the rewrite also needs to acknowledge the variable C which way say we'll refer back to but don't as yet.

    ax.loglog(xvals, yvals, color='grey')
```

> where the maximum word frequency rank corresponds to \(a\)
> in the general form of a power law and \(-\beta + 1\)
> corresponds to \(b\).

Reviewer: ```suggestion
where the maximum word frequency rank corresponds to intercept
in our linear form, and \(-\beta + 1\) the slope.
```
